network/gossip/libp2p/peerManager_test.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/model/flow/order"
+	"github.com/onflow/flow-go/network/gossip/libp2p/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type PeerManagerTestSuite struct {
+	suite.Suite
+	log zerolog.Logger
+	ctx context.Context
+}
+
+func TestPeerManagerTestSuite(t *testing.T) {
+	suite.Run(t, new(PeerManagerTestSuite))
+}
+
+func (ts *PeerManagerTestSuite) SetupTest() {
+	ts.log = ts.log.Output(zerolog.ConsoleWriter{Out: os.Stderr}).With().Caller().Logger()
+	ts.ctx = context.Background()
+	err := InitializePeerInfoCache()
+	assert.NoError(ts.T(), err)
+}
+
+// TestUpdatePeers tests that updatePeers calls the connector with the expected list of ids to connect and disconnect
+// from
+func (ts *PeerManagerTestSuite) TestUpdatePeers() {
+
+	// create some test ids
+	currentIDs := unittest.IdentityListFixture(10)
+
+	// setup a ID provider callback to return currentIDs
+	idProvider := func() (flow.IdentityList, error) {
+		return currentIDs, nil
+	}
+
+	// track IDs that should be disconnected
+	var extraIDs flow.IdentityList
+
+	// create the connector mock to check ids requested for connect and disconnect
+	connector := new(mock.Connector)
+	connector.On("ConnectPeers", ts.ctx, mock2.AnythingOfType("flow.IdentityList")).
+		Run(func(args mock2.Arguments) {
+			idArg := args[1].(flow.IdentityList)
+			assertListsEqual(ts.T(), currentIDs, idArg)
+		}).
+		Return(nil)
+	connector.On("DisconnectPeers", ts.ctx, mock2.AnythingOfType("flow.IdentityList")).
+		Run(func(args mock2.Arguments) {
+			idArg := args[1].(flow.IdentityList)
+			assertListsEqual(ts.T(), extraIDs, idArg)
+			// assert that ids passed to disconnect have no id in common with those passed to connect
+			assertListsDisjoint(ts.T(), currentIDs, extraIDs)
+		}).
+		Return(nil)
+
+	// create the peer manager (but don't start it)
+	pm := NewPeerManager(ts.ctx, ts.log, idProvider, connector)
+
+	// very first call to updatepeer
+	ts.Run("updatePeers only connects to all peers the first time", func() {
+
+		pm.updatePeers()
+
+		connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 1)
+		connector.AssertNotCalled(ts.T(), "DisconnectPeers")
+	})
+
+	// a subsequent call to updatepeer should request a connect to existing ids and new ids
+	ts.Run("updatePeers connects to old and new peers", func() {
+		// create a new id
+		newIDs := unittest.IdentityListFixture(1)
+		currentIDs = append(currentIDs, newIDs...)
+
+		pm.updatePeers()
+
+		connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 2)
+		connector.AssertNotCalled(ts.T(), "DisconnectPeers")
+	})
+
+	// when ids are excluded, they should be requested to be disconnected
+	ts.Run("updatePeers disconnects from extra peers", func() {
+		// delete an id
+		extraIDs = currentIDs.Sample(1)
+		currentIDs = currentIDs.Filter(filter.Not(filter.In(extraIDs)))
+
+		pm.updatePeers()
+
+		connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 3)
```

Comment:
these calls are cumulative across the test runs right? A comment to that effect would be nice since that's a bit different from most of our tests.

Reply:
yes they are cumulative. Will add comment.

Reply:
updated comment